File: T1.SqlSharp/T1.SqlSharp/Expressions/SqlConstraintForeignKey.cs
using System.Text;
using T1.SqlSharp.Extensions;

namespace T1.SqlSharp.Expressions;

public class SqlConstraintForeignKey : ISqlConstraint
{
    public SqlType SqlType { get; }= SqlType.TableForeignKey;
    public TextSpan Span { get; set; } = new();
    public string ConstraintName { get; set; } = string.Empty;
    public List<SqlConstraintColumn> Columns { get; set; } = [];
    public string ReferencedTableName { get; set; } = string.Empty;
    public string RefColumn { get; set; } = string.Empty; 
    public ReferentialAction OnDeleteAction { get; set; } = ReferentialAction.NoAction;
    public ReferentialAction OnUpdateAction { get; set; } = ReferentialAction.NoAction;
    public bool NotForReplication { get; set; }

    public string ToSql()
    {
        var sql = new StringBuilder();
        if (!string.IsNullOrEmpty(ConstraintName))
        {
            sql.Append($"CONSTRAINT {ConstraintName} ");
        }

        sql.Append("(");
        sql.Append(string.Join(", ", Columns.Select(c => c.ToSql())));
        sql.Append(")");
        
        sql.Append($" REFERENCES {ReferencedTableName}");
        if (!string.IsNullOrEmpty(RefColumn))
        {
            sql.Append($"({RefColumn})");
        }
        if (OnDeleteAction != ReferentialAction.NoAction)
        {
            sql.Append($" ON DELETE {OnDeleteAction.ToSql()}");
        }
        if (OnUpdateAction != ReferentialAction.NoAction)
        {
            sql.Append($" ON UPDATE {OnUpdateAction.ToSql()}");
        }
        if(NotForReplication)
        {
            sql.Append(" NOT FOR REPLICATION");
        }
        return sql.ToString();
    }
}

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlConstraintForeignKey.ts
import { SqlType } from './SqlType';
import { TextSpan } from '../StringParser';
import { ISqlConstraint, ISqlExpression, SqlVisitor } from './ISqlExpression';
import { SqlConstraintColumn } from './SqlConstraintColumn';
import { ReferentialAction } from './ReferentialAction';

export class SqlConstraintForeignKey implements ISqlConstraint {
    SqlType: SqlType = SqlType.TableForeignKey;
    Span: TextSpan = new TextSpan();
    ConstraintName: string = '';
    Columns: SqlConstraintColumn[] = [];
    ReferencedTableName: string = '';
    RefColumn: string = '';
    OnDeleteAction: ReferentialAction = ReferentialAction.NoAction;
    OnUpdateAction: ReferentialAction = ReferentialAction.NoAction;
    NotForReplication: boolean = false;

    Accept(visitor: SqlVisitor): void {
        visitor.Visit_ConstraintForeignKey(this);
    }

    ToSql(): string {
        let sql = '';
        if (this.ConstraintName) {
            sql += `CONSTRAINT ${this.ConstraintName} `;
        }
        sql += '(' + this.Columns.map(c => c.ToSql()).join(', ') + ')';
        sql += ` REFERENCES ${this.ReferencedTableName}`;
        if (this.RefColumn) {
            sql += `(${this.RefColumn})`;
        }
        if (this.OnDeleteAction !== ReferentialAction.NoAction) {
            sql += ` ON DELETE ${this.OnDeleteAction}`;
        }
        if (this.OnUpdateAction !== ReferentialAction.NoAction) {
            sql += ` ON UPDATE ${this.OnUpdateAction}`;
        }
        if (this.NotForReplication) {
            sql += ' NOT FOR REPLICATION';
        }
        return sql;
    }
} 